src/backend/distributed/operations/repair_shards.c

```diff
@@ -158,6 +178,125 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * ColocationSizeInBytes returns the size in bytes of a set of colocated tables.
+ */
+int64
+ColocationSizeInBytes(List *colocatedShardList, char *workerNodeName, uint32
+					  workerNodePort)
+{
+	uint32 connectionFlag = 0;
+
+	/* we skip child tables of a partitioned table if this boolean variable is true */
+	bool optimizePartitionCalculations = true;
+	StringInfo tableSizeQuery = GenerateSizeQueryOnMultiplePlacements(colocatedShardList,
+																	  TOTAL_RELATION_SIZE,
+																	  optimizePartitionCalculations);
+
+	MultiConnection *connection = GetNodeConnection(connectionFlag, workerNodeName,
+													workerNodePort);
+	PGresult *result = NULL;
+	int queryResult = ExecuteOptionalRemoteCommand(connection, tableSizeQuery->data,
+												   &result);
+
+	if (queryResult != RESPONSE_OKAY)
+	{
+		ereport(ERROR, (errcode(ERRCODE_CONNECTION_FAILURE),
+						errmsg("cannot get the size because of a connection error")));
+	}
+
+	List *sizeList = ReadFirstColumnAsText(result);
+	if (list_length(sizeList) != 1)
+	{
+		ereport(ERROR, (errmsg(
+							"received wrong number of rows from worker, expected 1 received %d",
+							list_length(sizeList))));
+	}
+
+	StringInfo tableSizeStringInfo = (StringInfo) linitial(sizeList);
+	char *tableSizeString = tableSizeStringInfo->data;
+	uint64 tableSize = SafeStringToUint64(tableSizeString);
+
+	bool raiseErrors = true;
+	PQclear(result);
+	ClearResults(connection, raiseErrors);
+
+	return tableSize;
+}
+
+
+/*
+ * CheckSpaceConstraints checks there is enough space to place the colocation
+ * on the node that the connection is connected to.
+ */
+static
+DeferredErrorMessage *
+CheckSpaceConstraints(MultiConnection *connection, int64 colocationSizeInBytes)
+{
+	if (!CheckAvailableSpaceBeforeMove)
```

Comment:
I expected to see this check at the start of `EnsureEnoughDiskSpaceForShardMove`. Is there a reason for us defer it until here?